src/command/shim.rs

```diff
@@ -100,38 +108,41 @@ Options:
     }
 }
 
-fn list(session: &mut Session, verbose: bool) -> Fallible<bool> {
+fn list(session: &Session, verbose: bool) -> Fallible<bool> {
     path::shim_dir()
         .and_then(|shim_dir| fs::read_dir(shim_dir).unknown())
         .map(|files| {
-            files
-                .map(|file| {
-                    file.and_then(|f| {
-                        f.path().file_name().map_or(Ok(false), |shim_name| {
-                            if verbose {
-                                match resolve_shim(session, &shim_name) {
-                                    Ok(shim_info) => {
-                                        println!("{} -> {}", shim_name.to_string_lossy(), shim_info)
-                                    },
-                                    Err(err) => {
-                                        style::display_error(style::ErrorContext::Notion, &err);
-                                        return Ok(false);
-                                    },
-                                }
-                            } else {
-                                println!("{}", shim_name.to_string_lossy());
-                            }
-                            Ok(true)
-                        })
-                    })
-                })
-                .collect::<Vec<_>>()
-                .iter()
-                // return false if anything failed
-                .all(|ref result| result.as_ref().ok() == Some(&true))
+            files.map(|file| {
+                file.and_then(|f| print_file_info(f, session, verbose))
+            })
+            .collect::<Vec<_>>()
+            .iter()
+            // return false if anything failed
+            .all(|ref result| result.as_ref().ok() == Some(&true))
         })
 }
 
+fn print_file_info(
```

Comment:
Here's a simplified version of this function:

```rust
fn print_file_info(file: fs::DirEntry, session: &Session, verbose: bool) -> Fallible<()> {
    let shim_name = file.file_name();

    if verbose {
        let shim_info = resolve_shim(session, &shim_name)?;
        println!("{} -> {}", shim_name.to_string_lossy(), shim_info);
    } else {
        println!("{}", shim_name.to_string_lossy());
    }

    Ok(())
}
```

Reply:
So much better!